Lib/_ldap.pyi

```diff
@@ -0,0 +1,406 @@
+RES_ADD: int
+RES_ANY: int
+RES_BIND: int
+RES_COMPARE: int
+RES_DELETE: int
+RES_EXTENDED: int
+RES_INTERMEDIATE: int
+RES_MODIFY: int
+RES_MODRDN: int
+RES_SEARCH_ENTRY: int
+RES_SEARCH_REFERENCE: int
+RES_SEARCH_RESULT: int
+RES_UNSOLICITED: int
+SASL_AUTOMATIC: int
+SASL_AVAIL: int
+SASL_INTERACTIVE: int
+SASL_QUIET: int
+SCOPE_BASE: int
+SCOPE_ONELEVEL: int
+SCOPE_SUBORDINATE: int
+SCOPE_SUBTREE: int
+SYNC_INFO: str
+TAG_CONTROLS: int
+TAG_EXOP_REQ_OID: int
+TAG_EXOP_REQ_VALUE: int
+TAG_EXOP_RES_OID: int
+TAG_EXOP_RES_VALUE: int
+TAG_LDAPCRED: int
+TAG_LDAPDN: int
+TAG_MESSAGE: int
+TAG_MSGID: int
+TAG_NEWSUPERIOR: int
+TAG_REFERRAL: int
+TAG_SASL_RES_CREDS: int
+TLS_AVAIL: int
+URL_ERR_BADSCOPE: int
+URL_ERR_MEM: int
+VENDOR_VERSION: int
+VERSION: int
+VERSION1: int
+VERSION2: int
+VERSION3: int
+VERSION_MAX: int
+VERSION_MIN: int
+
+class ADMINLIMIT_EXCEEDED(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class AFFECTS_MULTIPLE_DSAS(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class ALIAS_DEREF_PROBLEM(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class ALIAS_PROBLEM(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class ALREADY_EXISTS(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class ASSERTION_FAILED(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class AUTH_METHOD_NOT_SUPPORTED(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class AUTH_UNKNOWN(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class BUSY(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CANCELLED(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CANNOT_CANCEL(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CLIENT_LOOP(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class COMPARE_FALSE(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class COMPARE_TRUE(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CONFIDENTIALITY_REQUIRED(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CONNECT_ERROR(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CONSTRAINT_VIOLATION(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class CONTROL_NOT_FOUND(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class DECODING_ERROR(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class ENCODING_ERROR(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class FILTER_ERROR(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class INAPPROPRIATE_AUTH(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class INAPPROPRIATE_MATCHING(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class INSUFFICIENT_ACCESS(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class INVALID_CREDENTIALS(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class INVALID_DN_SYNTAX(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class INVALID_SYNTAX(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class IS_LEAF(ldap.LDAPError):
+    errnum: ClassVar[int] = ...
+
+class LDAPError(Exception): ...
```

Comment:
Mypy does not consider these error classes and constants to be re-exported from the `ldap` module.

So when I use `import ldap; ldap.SERVER_DOWN`, it causes "error: Module has no attribute "SERVER_DOWN" [attr-defined]". Same with constants such as `SCOPE_SUBTREE`, `MOD_REPLACE` etc.

`Lib/ldap/__init__.py` contains `from _ldap import *`. I'm a bit confused though, https://typing.readthedocs.io/en/latest/source/stubs.html#imports lists `from Y import *` as being a re-export. Not sure what is happening.

Reply:
> ldap.SERVER_DOWN

Actually, I realised that adding `py.typed` and `_ldap.pyi` to `MANIFEST.in` wasn't sufficient (and if you check the `python_ldap-*` dir in `site-packages` (presumably in your virtual environment), the files aren't there.

I fixed that locally by adding `extra_files = .:Lib/_ldap.pyi` to the `[_ldap]` section of `setup.cfg` and adding `package_data = {'ldap': ['py.typed'],},` towards the bottom of `setup.py`. The files now get installed, but `mypy` still isn't able to find the `_ldap.pyi` stub... 😕

Reply:
I already tried that (and also tried creating an empty top-level `__init__.py` file)...still no dice

Reply:
Here some solutions are discussed:

* https://github.com/python/typing/issues/1333

One approach is to convert the `_ldap` module to a package first.

But I wonder if including an `_ldap-stubs` directory with `__init__.pyi` would also solve the issue.

Reply:
> Here some solutions are discussed, but they're not very nice:
> 
> * [Type stubs for single-file top-level modules python/typing#1333](https://github.com/python/typing/issues/1333)

Hm, good link....but no real solutions IMHO.

> One approach is to convert the `_ldap` module to a package first.
> 
> But I wonder if including an `_ldap-stubs` directory with `__init__.pyi` would also solve the issue.

I tried manually moving `_ldap.*` into the `.../lib/python3.x/site_packages/python_ldap-*/ldap` directory and changing a few imports, and that actually seems to work. Would mean that `_ldap` is instead `ldap._ldap`, but given that it's an internal implementation detail, that might actually be ok...I'll work on a patch to see what it might look like...

Reply:
> > Here some solutions are discussed, but they're not very nice:
> > 
> > * [Type stubs for single-file top-level modules python/typing#1333](https://github.com/python/typing/issues/1333)
> 
> Hm, good link....but no real solutions IMHO.
> 
> > One approach is to convert the `_ldap` module to a package first.
> > But I wonder if including an `_ldap-stubs` directory with `__init__.pyi` would also solve the issue.
> 
> I tried manually moving `_ldap.*` into the `.../lib/python3.x/site_packages/python_ldap-*/ldap` directory and changing a few imports, and that actually seems to work. Would mean that `_ldap` is instead `ldap._ldap`, but given that it's an internal implementation detail, that might actually be ok...I'll work on a patch to see what it might look like...

Ok, I've managed to reflect that in the build system as well. I'll go with that solution for now (or until someone smarter than me comes up with a better solution).